ai_powered_qa/components/agent.py

```diff
@@ -153,23 +153,26 @@ def _get_messages_for_completion(
         if user_prompt:
             total_tokens += count_tokens(user_prompt, model)
 
-        for i in range(len(self.history)):
+        messages_to_add = []
+        i = 0
+
+        while i < len(self.history):
             history_item = self.history[-i - 1]
-            content_length = count_tokens(history_item["content"], model)
+            messages_to_add.insert(0, history_item)
+            content_length = count_tokens(yaml.dump(history_item), model)
 
-            if "tool_calls" in history_item:
-                content_length += count_tokens(
-                    yaml.dump(history_item["tool_calls"]), model
-                )
-                content_length += self._count_tokens_for_tool_responses(
-                    self.history[-i : -i - len(history_item["tool_calls"])], model
-                )
-                i += len(history_item["tool_calls"])
+            while history_item["role"] == "tool":
```

Comment:
Going from the back of the history, it should always be 1 or more messages from `tool`, followed by a single `assistant` message which triggered the tool calls